collector/zoneinfo_linux.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/go-kit/kit/log"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs"
+)
+
+const zoneinfoSubsystem = "zoneinfo"
+
+type zoneinfoCollector struct {
+	gaugeMetricDescs   map[string]*prometheus.Desc
+	counterMetricDescs map[string]*prometheus.Desc
+	logger             log.Logger
+	fs                 procfs.FS
+}
+
+func init() {
+	registerCollector("zoneinfo", defaultDisabled, NewZoneinfoCollector)
+}
+
+// NewZoneinfoCollector returns a new Collector exposing zone stats.
+func NewZoneinfoCollector(logger log.Logger) (Collector, error) {
+	fs, err := procfs.NewFS(*procPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open procfs: %w", err)
+	}
+	return &zoneinfoCollector{
+		gaugeMetricDescs:   createGaugeMetricDescriptions(),
+		counterMetricDescs: createCounterMetricDescriptions(),
+		logger:             logger,
+		fs:                 fs,
+	}, nil
+}
+
+func (c *zoneinfoCollector) Update(ch chan<- prometheus.Metric) error {
+	metrics, err := c.fs.Zoneinfo()
+	if err != nil {
+		return fmt.Errorf("couldn't get zoneinfo: %w", err)
+	}
+	for _, metric := range metrics {
+		node := metric.Node
+		zone := metric.Zone
+		metricStruct := reflect.ValueOf(metric)
+		typeOfMetricStruct := metricStruct.Type()
+		for i := 0; i < metricStruct.NumField(); i++ {
+			value := reflect.Indirect(metricStruct.Field(i))
+			if value.Kind() != reflect.Int64 {
+				continue
+			}
+			metricName := typeOfMetricStruct.Field(i).Name
+			desc, ok := c.gaugeMetricDescs[metricName]
+			metricType := prometheus.GaugeValue
+			if !ok {
+				desc = c.counterMetricDescs[metricName]
+				metricType = prometheus.CounterValue
+			}
+			ch <- prometheus.MustNewConstMetric(desc, metricType,
+				float64(reflect.Indirect(metricStruct.Field(i)).Int()),
+				node, zone)
+		}
+		for i, value := range metric.Protection {
+			metricName := fmt.Sprintf("protection_%d", i)
+			desc, ok := c.gaugeMetricDescs[metricName]
+			if !ok {
+				desc = prometheus.NewDesc(
+					prometheus.BuildFQName(namespace, zoneinfoSubsystem, metricName),
+					fmt.Sprintf("Protection array %d. field", i),
+					[]string{"node", "zone"}, nil)
+				c.gaugeMetricDescs[metricName] = desc
+			}
+			ch <- prometheus.MustNewConstMetric(desc, prometheus.GaugeValue,
+				float64(*value), node, zone)
+		}
```

Comment:
Can you explain what these metric represent? Or how they would be used in a query?

Reply:
This metric shows, whether or not zone should be used for page allocation. This is explained [here](https://android.googlesource.com/kernel/common/+/bcmdhd-3.10/Documentation/sysctl/vm.txt#288)

Reply:
I'm mostly wondering how they would be used in practice. What PromQL queries would you use them for? Can you share some examples?

Reply:
I can't provide any real use case for it, my idea was, that I'll leave it for the user, but if you find this metric problematic, I can remove it or make it like `protection(zone= ith field)` (make each field of array label in metric instead of whole metric)

Reply:
I guess this is fine for now, since this is a default disabled collector.

Reply:
My objection is that if we can't figure out a good use case for it, it's probably best to leave it out since it's not something you can use for monitoring.